### PR TITLE
refine build-in stages

### PR DESCRIPTION
**Pull request by Hawstein**
* change InHandler to avoid unnecessary condition check
* remove unnecessary local variable
* fix typo

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

**Comment by ktoso** [comment]
PLS WHITELIST

**Comment by Hawstein** [comment]
I guess this PR needs a `PLS BUILD` 😄

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 1]
```diff
@@ -279,12 +279,11 @@ final case class Take[T](count: Long) extends SimpleLinearGraphStage[T] {
...
     override def onPush(): Unit = {
-      val leftBefore = left
-      if (leftBefore >= 1) {
-        left = leftBefore - 1
+      if (left > 0) {
         push(out, grab(in))
+        left -= 1
       }
-      if (leftBefore <= 1) completeStage()
+      if (left <= 0) completeStage()
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: could be left == 0, but this is also fine

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 2]
```diff
@@ -1588,21 +1591,19 @@ final class TakeWithin[T](val timeout: FiniteDuration) extends SimpleLinearGraph
 final class DropWithin[T](val timeout: FiniteDuration) extends SimpleLinearGraphStage[T] {
   override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) with InHandler with OutHandler {
 
-    private var allow = false
+    override def preStart(): Unit = scheduleOnce("DropWithinTimer", timeout)
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: define the schedule key as a case object in companion
```scala
object DropWithin {
  private case object DropWithinTimer
}
```

Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: hey, where was the `scheduleOnce` in the old code? was it a bug? don't we have test for this stage?

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: L1605. It was just moved up here.

**Review by patriknw — APPROVED** [review]
LGTM

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 3]
```diff
@@ -311,7 +310,13 @@ final case class Drop[T](count: Long) extends SimpleLinearGraphStage[T] {
       if (left > 0) {
         left -= 1
         pull(in)
-      } else push(out, grab(in))
+      } else {
+        push(out, grab(in))
+        // After dropping N elements, the following behavior is just pushing
+        setHandler(in, new InHandler {
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: hm, this is a tradeoff between an extra allocation and a branch which is totally predictable. I vote for the branch here.

**Review by drewhk — APPROVED** [review]

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 2 above)
Comment (jrudolph) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Does that actually need the timer at all? Wouldn't it be simpler (and cheaper) to check the end of drop period in `onPush`?

(This also points to the other issue that for this (and maybe other) timing-related stages there could be alternative implementation that would eagerly pull and drop elements during the initial drop phase.)

**Comment by patriknw** [comment]
@Hawstein There seems to be a few comments here. Can you update, or what is the plan?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 3 above)
Comment (Hawstein) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: You are right. Although I prefer the *branch elimination* version, the original version is OK to me.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 2 above)
Comment (Hawstein) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: @jrudolph Good idea. So here I am going to remove the timer. The TimerGraphStageLogic refactorization should be put in another PR.

**Comment by Hawstein** [comment]
@patriknw @drewhk @jrudolph Have replied the comments and updated the PR.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 4]
```diff
@@ -1608,23 +1605,27 @@ final class TakeWithin[T](val timeout: FiniteDuration) extends SimpleLinearGraph
...
+      }
     }
 
     def onPull(): Unit = pull(in)
 
-    setHandler(in, this)
-    setHandler(out, this)
-
-    final override protected def onTimer(key: Any): Unit = allow = true
+    setHandlers(in, out, this)
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: ok refactoring

**Review by ktoso — APPROVED** [review]
LGTM, thanks!

**Comment by ktoso** [comment]
Was there a ticket for this?

**Comment by Hawstein** [comment]
@ktoso Nope, this is just some very little modification. So I didn't create a ticket for it.
